src/CodeGen_GPU_Dev.cpp

```diff
@@ -149,38 +149,76 @@ void CodeGen_GPU_C::visit(const Shuffle *op) {
             internal_assert(op->vectors[0].type() == op->vectors[i].type());
         }
         internal_assert(op->type.lanes() == (int)op->indices.size());
-        const int max_index = (int)(op->vectors[0].type().lanes() * op->vectors.size());
+        std::vector<int> vector_first_index;
+        int max_index = 0;
+        for (const Expr &v : op->vectors) {
+            vector_first_index.push_back(max_index);
+            max_index += v.type().lanes();
+        }
         for (int i : op->indices) {
             internal_assert(i >= 0 && i < max_index);
         }
 
         std::vector<std::string> vecs;
         for (const Expr &v : op->vectors) {
-            vecs.push_back(print_expr(v));
+            std::string expr = print_expr(v);
+            internal_assert(!expr.empty()) << "Expression did not serialize.";
+            vecs.push_back(std::move(expr));
         }
 
         std::string src = vecs[0];
         std::ostringstream rhs;
         std::string storage_name = unique_name('_');
-        if (vector_declaration_style == VectorDeclarationStyle::OpenCLSyntax) {
+        switch (vector_declaration_style) {
+        case VectorDeclarationStyle::OpenCLSyntax:
             rhs << "(" << print_type(op->type) << ")(";
-        } else if (vector_declaration_style == VectorDeclarationStyle::WGSLSyntax) {
+            break;
+        case VectorDeclarationStyle::WGSLSyntax:
             rhs << print_type(op->type) << "(";
-        } else {
+            break;
+        case VectorDeclarationStyle::CLikeSyntax:
             rhs << "{";
+            break;
         }
+        int elem_num = 0;
         for (int i : op->indices) {
-            rhs << vecs[i];
-            if (i < (int)(op->indices.size() - 1)) {
+            size_t vector_idx;
+            int lane_idx = -1;
+            for (vector_idx = 0; vector_idx < op->vectors.size(); ++vector_idx) {
+                if (i >= vector_first_index[vector_idx] && i < vector_first_index[vector_idx] + op->vectors[vector_idx].type().lanes()) {
+                    lane_idx = i - vector_first_index[vector_idx];
+                    break;
+                }
+            }
+            internal_assert(lane_idx != -1) << "Shuffle lane index not found.";
+            internal_assert(vector_idx < op->vectors.size()) << "Shuffle vector index not found.";
```

Comment:
nit: these error messages should also print more info, e.g.
```
internal_assert(lane_idx != -1) << "Shuffle lane index not found: " << i;
internal_assert(vector_idx < op->vectors.size()) << "Shuffle vector index: " << i << " not found.";
```

Reply:
They are internal asserts that should never trigger. I don't think it's that important to add in those details.

Reply:
Implemented your suggestion either way.